x/withdraw/keeper/ibc_callbacks.go

```diff
@@ -0,0 +1,182 @@
+		return ack
+	}
+
+	// unmarshal packet data to obtain the sender and recipient
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		err = sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data")
+		return channeltypes.NewErrorAcknowledgement(err.Error())
+	}
+
+	// validate the sender bech32 address from the counterparty chain
+	bech32Prefix := strings.Split(data.Sender, "1")[0]
+	if bech32Prefix == data.Sender {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender: %s", data.Sender).Error(),
+		)
+	}
+
+	senderBz, err := sdk.GetFromBech32(data.Sender, bech32Prefix)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender %s, %s", data.Sender, err.Error()).Error(),
+		)
+	}
+
+	// change the bech32 human readable prefix (HRP) of the sender to `evmos1`
+	sender := sdk.AccAddress(senderBz)
+
+	// obtain the evmos recipient address
+	recipient, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid receiver address %s", err.Error()).Error(),
+		)
+	}
+
+	// case 1: sender ≠ recipient.
+	// Withdraw is only possible for addresses in which the sender = recipient.
+	// Continue to the next IBC middleware by returning the original ACK.
+	if !sender.Equals(recipient) {
+		return ack
+	}
+
+	// get the recipient account
+	account := k.accountKeeper.GetAccount(ctx, recipient)
+
+	// Case 2. recipient pubkey is a supported key (eth_secp256k1, amino multisig, ed25519)
+	// ==> Continue and return success ACK as the funds are not stuck on chain
+	if account != nil &&
+		account.GetPubKey() != nil &&
+		evmos.IsSupportedKey(account.GetPubKey()) {
+		return ack
+	}
+
+	// NOTE: Since destination channel is authorized and not from an EVM chain, we know that
+	// only secp256k1 keys are supported in the source chain. This means that we can now
+	// initiate the withdraw logic
+
+	// transfer the balance back to the sender address
+	destPort := packet.DestinationPort
+	destChannel := packet.DestinationChannel
+	balances := sdk.Coins{}
+
+	k.bankKeeper.IterateAccountBalances(ctx, recipient, func(coin sdk.Coin) (stop bool) {
+		if coin.IsZero() {
+			// continue
+			return false
+		}
+
+		switch strings.HasPrefix(coin.Denom, "ibc/") {
+		case true:
+			// IBC vouchers, obtain the source port and channel from the denom path
+			destPort, destChannel, err = k.GetIBCDenomSelfIdentifiers(ctx, coin.Denom, data.Sender)
+		default:
+			// Native tokens, use the source port and channel to transfer the EVMOS and
+			// other converted ERC20 coin denoms to the authorized source chain
+			destPort = packet.DestinationPort
+			destChannel = packet.DestinationChannel
+		}
+
+		if err != nil {
```

Comment:
so first transfer from authorized source chain:
- sends back IBC tokens which originated from the source chain
- sends over all Evmos native tokens

second transfer from a different authorized source chain:
- presumably only sends back IBC tokens which originated from the source chain